File: pkg/otel/metrics/otlp/related_data.go
/*
 * Copyright The OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package otlp

import (
	colarspb "github.com/f5/otel-arrow-adapter/api/experimental/arrow/v1"
	"github.com/f5/otel-arrow-adapter/pkg/otel"
	"github.com/f5/otel-arrow-adapter/pkg/otel/common/otlp"
	"github.com/f5/otel-arrow-adapter/pkg/record_message"
	"github.com/f5/otel-arrow-adapter/pkg/werror"
)

type (
	RelatedData struct {
		MetricID uint16

		// Attributes stores
		ResAttrMapStore                *otlp.Attributes16Store
		ScopeAttrMapStore              *otlp.Attributes16Store
		NumberDPAttrsStore             *otlp.Attributes32Store
		SummaryAttrsStore              *otlp.Attributes32Store
		HistogramAttrsStore            *otlp.Attributes32Store
		ExpHistogramAttrsStore         *otlp.Attributes32Store
		NumberDPExemplarAttrsStore     *otlp.Attributes32Store
		HistogramExemplarAttrsStore    *otlp.Attributes32Store
		ExpHistogramExemplarAttrsStore *otlp.Attributes32Store

		// Metric stores
		NumberDataPointsStore     *NumberDataPointsStore
		SummaryDataPointsStore    *SummaryDataPointsStore
		HistogramDataPointsStore  *HistogramDataPointsStore
		EHistogramDataPointsStore *EHistogramDataPointsStore

		// Exemplar stores
		NumberDataPointExemplarsStore     *ExemplarsStore
		HistogramDataPointExemplarsStore  *ExemplarsStore
		EHistogramDataPointExemplarsStore *ExemplarsStore
	}
)

func NewRelatedData() *RelatedData {
	return &RelatedData{
		ResAttrMapStore:                otlp.NewAttributes16Store(),
		ScopeAttrMapStore:              otlp.NewAttributes16Store(),
		NumberDPAttrsStore:             otlp.NewAttributes32Store(),
		SummaryAttrsStore:              otlp.NewAttributes32Store(),
		HistogramAttrsStore:            otlp.NewAttributes32Store(),
		ExpHistogramAttrsStore:         otlp.NewAttributes32Store(),
		NumberDPExemplarAttrsStore:     otlp.NewAttributes32Store(),
		HistogramExemplarAttrsStore:    otlp.NewAttributes32Store(),
		ExpHistogramExemplarAttrsStore: otlp.NewAttributes32Store(),

		NumberDataPointsStore:     NewNumberDataPointsStore(),
		SummaryDataPointsStore:    NewSummaryDataPointsStore(),
		HistogramDataPointsStore:  NewHistogramDataPointsStore(),
		EHistogramDataPointsStore: NewEHistogramDataPointsStore(),

		NumberDataPointExemplarsStore:     NewExemplarsStore(),
		HistogramDataPointExemplarsStore:  NewExemplarsStore(),
		EHistogramDataPointExemplarsStore: NewExemplarsStore(),
	}
}

func (r *RelatedData) MetricIDFromDelta(delta uint16) uint16 {
	r.MetricID += delta
	return r.MetricID
}

func RelatedDataFrom(records []*record_message.RecordMessage) (relatedData *RelatedData, metricsRecord *record_message.RecordMessage, err error) {
	var numberDPRec *record_message.RecordMessage
	var summaryDPRec *record_message.RecordMessage
	var histogramDPRec *record_message.RecordMessage
	var expHistogramDPRec *record_message.RecordMessage
	var numberDBExRec *record_message.RecordMessage
	var histogramDBExRec *record_message.RecordMessage
	var expHistogramDBExRec *record_message.RecordMessage

	relatedData = NewRelatedData()

	for _, record := range records {
		switch record.PayloadType() {
		case colarspb.ArrowPayloadType_RESOURCE_ATTRS:
			err = otlp.Attributes16StoreFrom(record.Record(), relatedData.ResAttrMapStore)
			if err != nil {
				return nil, nil, werror.Wrap(err)
			}
		case colarspb.ArrowPayloadType_SCOPE_ATTRS:
			err = otlp.Attributes16StoreFrom(record.Record(), relatedData.ScopeAttrMapStore)
			if err != nil {
				return nil, nil, werror.Wrap(err)
			}
		case colarspb.ArrowPayloadType_NUMBER_DP_ATTRS:
			err = otlp.Attributes32StoreFrom(record.Record(), relatedData.NumberDPAttrsStore)
			if err != nil {
				return nil, nil, werror.Wrap(err)
			}
		case colarspb.ArrowPayloadType_SUMMARY_DP_ATTRS:
			err = otlp.Attributes32StoreFrom(record.Record(), relatedData.SummaryAttrsStore)
			if err != nil {
				return nil, nil, werror.Wrap(err)
			}
		case colarspb.ArrowPayloadType_HISTOGRAM_DP_ATTRS:
			err = otlp.Attributes32StoreFrom(record.Record(), relatedData.HistogramAttrsStore)
			if err != nil {
				return nil, nil, werror.Wrap(err)
			}
		case colarspb.ArrowPayloadType_EXP_HISTOGRAM_DP_ATTRS:
			err = otlp.Attributes32StoreFrom(record.Record(), relatedData.ExpHistogramAttrsStore)
			if err != nil {
				return nil, nil, werror.Wrap(err)
			}
		case colarspb.ArrowPayloadType_NUMBER_DATA_POINTS:
			if numberDPRec != nil {
				return nil, nil, werror.Wrap(otel.ErrDuplicatePayloadType)
			}
			numberDPRec = record
		case colarspb.ArrowPayloadType_SUMMARY_DATA_POINTS:
			if summaryDPRec != nil {
				return nil, nil, werror.Wrap(otel.ErrDuplicatePayloadType)
			}
			summaryDPRec = record
		case colarspb.ArrowPayloadType_HISTOGRAM_DATA_POINTS:
			if histogramDPRec != nil {
				return nil, nil, werror.Wrap(otel.ErrDuplicatePayloadType)
			}
			histogramDPRec = record
		case colarspb.ArrowPayloadType_EXP_HISTOGRAM_DATA_POINTS:
			if expHistogramDPRec != nil {
				return nil, nil, werror.Wrap(otel.ErrDuplicatePayloadType)
			}
			expHistogramDPRec = record
		case colarspb.ArrowPayloadType_METRICS:
			if metricsRecord != nil {
				return nil, nil, werror.Wrap(otel.ErrDuplicatePayloadType)
			}
			metricsRecord = record
		case colarspb.ArrowPayloadType_NUMBER_DP_EXEMPLARS:
			if numberDBExRec != nil {
				return nil, nil, werror.Wrap(otel.ErrDuplicatePayloadType)
			}
			numberDBExRec = record
		case colarspb.ArrowPayloadType_HISTOGRAM_DP_EXEMPLARS:
			if histogramDBExRec != nil {
				return nil, nil, werror.Wrap(otel.ErrDuplicatePayloadType)
			}
			histogramDBExRec = record
		case colarspb.ArrowPayloadType_EXP_HISTOGRAM_DP_EXEMPLARS:
			if expHistogramDBExRec != nil {
				return nil, nil, werror.Wrap(otel.ErrDuplicatePayloadType)
			}
			expHistogramDBExRec = record
		case colarspb.ArrowPayloadType_NUMBER_DP_EXEMPLAR_ATTRS:
			err = otlp.Attributes32StoreFrom(record.Record(), relatedData.NumberDPExemplarAttrsStore)
			if err != nil {
				return nil, nil, werror.Wrap(err)
			}
		case colarspb.ArrowPayloadType_HISTOGRAM_DP_EXEMPLAR_ATTRS:
			err = otlp.Attributes32StoreFrom(record.Record(), relatedData.HistogramExemplarAttrsStore)
			if err != nil {
				return nil, nil, werror.Wrap(err)
			}
		case colarspb.ArrowPayloadType_EXP_HISTOGRAM_DP_EXEMPLAR_ATTRS:
			err = otlp.Attributes32StoreFrom(record.Record(), relatedData.ExpHistogramExemplarAttrsStore)
			if err != nil {
				return nil, nil, werror.Wrap(err)
			}
		default:
			return nil, nil, werror.Wrap(otel.UnknownPayloadType)
		}
	}

	// Process exemplar records
	if numberDBExRec != nil {
		relatedData.NumberDataPointExemplarsStore, err = ExemplarsStoreFrom(
			numberDBExRec.Record(),
			relatedData.NumberDPExemplarAttrsStore,
		)
		if err != nil {
			return nil, nil, werror.Wrap(err)
		}
	}

	if histogramDBExRec != nil {
		relatedData.HistogramDataPointExemplarsStore, err = ExemplarsStoreFrom(
			histogramDBExRec.Record(),
			relatedData.HistogramExemplarAttrsStore,
		)
		if err != nil {
			return nil, nil, werror.Wrap(err)
		}
	}

	if expHistogramDBExRec != nil {
		relatedData.EHistogramDataPointExemplarsStore, err = ExemplarsStoreFrom(
			expHistogramDBExRec.Record(),
			relatedData.ExpHistogramExemplarAttrsStore,
		)
		if err != nil {
			return nil, nil, werror.Wrap(err)
		}
	}

	// Process data point records
	if numberDPRec != nil {
		relatedData.NumberDataPointsStore, err = NumberDataPointsStoreFrom(
			numberDPRec.Record(),
			relatedData.NumberDataPointExemplarsStore,
			relatedData.NumberDPAttrsStore,
		)
		if err != nil {
			return nil, nil, werror.Wrap(err)
		}
	}

	if summaryDPRec != nil {
		relatedData.SummaryDataPointsStore, err = SummaryDataPointsStoreFrom(
			summaryDPRec.Record(),
			relatedData.SummaryAttrsStore,
		)
		if err != nil {
			return nil, nil, werror.Wrap(err)
		}
	}

	if histogramDPRec != nil {
		relatedData.HistogramDataPointsStore, err = HistogramDataPointsStoreFrom(
			histogramDPRec.Record(),
			relatedData.HistogramDataPointExemplarsStore,
			relatedData.HistogramAttrsStore,
		)
		if err != nil {
			return nil, nil, werror.Wrap(err)
		}
	}

	if expHistogramDPRec != nil {
		relatedData.EHistogramDataPointsStore, err = EHistogramDataPointsStoreFrom(
			expHistogramDPRec.Record(),
			relatedData.EHistogramDataPointExemplarsStore,
			relatedData.ExpHistogramAttrsStore,
		)
		if err != nil {
			return nil, nil, werror.Wrap(err)
		}
	}

	return
}

File: rust/otel-arrow-rust/src/otlp/metrics/related_data.rs
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::error;
use crate::otap::OtapBatch;
use crate::otlp::attributes::store::{Attribute16Store, Attribute32Store};
use crate::otlp::metrics::data_points::data_point_store::{
    EHistogramDataPointsStore, HistogramDataPointsStore, NumberDataPointsStore,
    SummaryDataPointsStore,
};
use crate::otlp::metrics::exemplar::ExemplarsStore;
use crate::proto::opentelemetry::arrow::v1::ArrowPayloadType;

#[derive(Default)]
pub struct RelatedData {
    pub(crate) metric_id: u16,

    // Resource attributes.
    pub(crate) res_attr_map_store: Attribute16Store,
    pub(crate) scope_attr_map_store: Attribute16Store,
    pub(crate) number_d_p_attrs_store: Attribute32Store,
    pub(crate) summary_attrs_store: Attribute32Store,
    pub(crate) histogram_attrs_store: Attribute32Store,
    pub(crate) exp_histogram_attrs_store: Attribute32Store,

    // Exemplars
    pub(crate) number_data_point_exemplars_store: ExemplarsStore,
    pub(crate) histogram_data_point_exemplars_store: ExemplarsStore,
    pub(crate) e_histogram_data_point_exemplars_store: ExemplarsStore,

    // Exemplar attributes store
    pub(crate) number_d_p_exemplar_attrs_store: Attribute32Store,
    pub(crate) histogram_exemplar_attrs_store: Attribute32Store,
    pub(crate) exp_histogram_exemplar_attrs_store: Attribute32Store,

    // Data points
    pub(crate) number_data_points_store: NumberDataPointsStore,
    pub(crate) summary_data_points_store: SummaryDataPointsStore,
    pub(crate) histogram_data_points_store: HistogramDataPointsStore,
    pub(crate) e_histogram_data_points_store: EHistogramDataPointsStore,
}

impl RelatedData {
    pub fn metric_id_from_delta(&mut self, delta: u16) -> u16 {
        self.metric_id += delta;
        self.metric_id
    }
}

impl TryFrom<&OtapBatch> for RelatedData {
    type Error = error::Error;

    fn try_from(otap_batch: &OtapBatch) -> error::Result<Self> {
        let mut related_data = RelatedData::default();

        if let Some(rb) = otap_batch.get(ArrowPayloadType::ResourceAttrs) {
            related_data.res_attr_map_store = Attribute16Store::try_from(rb)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::ScopeAttrs) {
            related_data.scope_attr_map_store = Attribute16Store::try_from(rb)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::NumberDpExemplarAttrs) {
            related_data.number_d_p_exemplar_attrs_store = Attribute32Store::try_from(rb)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::NumberDpExemplars) {
            related_data.number_data_point_exemplars_store =
                ExemplarsStore::try_from(rb, &mut related_data.number_d_p_exemplar_attrs_store)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::NumberDpAttrs) {
            related_data.number_d_p_attrs_store = Attribute32Store::try_from(rb)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::NumberDataPoints) {
            related_data.number_data_points_store = NumberDataPointsStore::from_record_batch(
                rb,
                &mut related_data.number_data_point_exemplars_store,
                &related_data.number_d_p_attrs_store,
            )?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::SummaryDpAttrs) {
            related_data.summary_attrs_store = Attribute32Store::try_from(rb)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::SummaryDataPoints) {
            related_data.summary_data_points_store = SummaryDataPointsStore::from_record_batch(
                rb,
                &mut related_data.summary_attrs_store,
            )?
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::HistogramDpExemplarAttrs) {
            related_data.histogram_exemplar_attrs_store = Attribute32Store::try_from(rb)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::HistogramDpExemplars) {
            related_data.histogram_data_point_exemplars_store =
                ExemplarsStore::try_from(rb, &mut related_data.histogram_exemplar_attrs_store)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::HistogramDataPoints) {
            related_data.histogram_data_points_store = HistogramDataPointsStore::from_record_batch(
                rb,
                &mut related_data.histogram_data_point_exemplars_store,
                &related_data.histogram_attrs_store,
            )?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::HistogramDpAttrs) {
            related_data.histogram_attrs_store = Attribute32Store::try_from(rb)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::ExpHistogramDpAttrs) {
            related_data.exp_histogram_attrs_store = Attribute32Store::try_from(rb)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::ExpHistogramDpExemplarAttrs) {
            related_data.exp_histogram_exemplar_attrs_store = Attribute32Store::try_from(rb)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::ExpHistogramDpExemplars) {
            related_data.e_histogram_data_point_exemplars_store =
                ExemplarsStore::try_from(rb, &mut related_data.exp_histogram_exemplar_attrs_store)?;
        }

        if let Some(rb) = otap_batch.get(ArrowPayloadType::ExpHistogramDataPoints) {
            related_data.e_histogram_data_points_store =
                EHistogramDataPointsStore::from_record_batch(
                    rb,
                    &mut related_data.e_histogram_data_point_exemplars_store,
                    &related_data.exp_histogram_attrs_store,
                )?;
        }

        Ok(related_data)
    }
}
